drivers/eeprom/eeprom.c

```diff
@@ -0,0 +1,170 @@
+/*
+ * Copyright (C) 2019 Otto-von-Guericke-Universität Magdeburg
+ *
+ * This file is subject to the terms and conditions of the GNU Lesser
+ * General Public License v2.1. See the file LICENSE in the top level
+ * directory for more details.
+ */
+
+/**
+ * @ingroup   drivers_eeprom
+ * @{
+ *
+ * @file
+ * @brief     Implementation for EEPROM driver API
+ *
+ * @author    Marian Buschsieweke <marian.buschsieweke@ovgu.de>
+ * @}
+ */
+
+#include <errno.h>
+#include <string.h>
+
+#include "log.h"
+#include "eeprom.h"
+#include "eeprom_driver.h"
+#include "eeprom_params.h"
+
+#define ENABLE_DEBUG    (0)
+#include "debug.h"
+
+#ifndef EEPROM_SET_BUF
+/**
+ * @brief   Stack allocated buffer to implement the set fallback function
+ */
+#define EEPROM_SET_BUF              (32)
+#endif
+
+static eeprom_off_t _sizes[EEPROM_NUMOF];
+
+/**
+ * @brief       Check if the given device is valid and initialized
+ *
+ * @param[in]   dev     The device to check
+ *
+ * @retval      0       The device is not valid
+ * @retval      1       The device is valid
+ */
+static inline int _is_dev_valid(eeprom_t dev)
+{
+    return ((dev < EEPROM_NUMOF) && (_sizes[dev])) ? 1 : 0;
+}
+
+/**
+ * @brief       Check if the given range is valid
+ *
+ * @param[in]   dev     The device the range refers to
+ * @param[in]   pos     Start of the range to check
+ * @param[in]   len     Length of the range to check
+ *
+ * @retval      0       The range is not valid
+ * @retval      1       The range is valid
+ */
+static inline int _is_range_valid(eeprom_t dev, eeprom_off_t pos, size_t len)
+{
+    /* We basically want to check for "pos + len < _sizes[dev]", but we have to
+     * be a bit careful to prevent integer overflows */
+    if (len > _sizes[dev]) {
+        return 0;
+    }
+
+    if (_sizes[dev] - len < pos) {
+        return 0;
+    }
+
+    return 1;
+}
+
+void eeprom_init(void)
+{
+    for (unsigned i = 0; i < EEPROM_NUMOF; i++) {
```

Comment:
As long as you have the dummy device, this would lead to crashes.
```suggestion
    for (unsigned i = 0; i < EEPROM_NUMOF - 1; i++) {
```

Reply:
The dummy is not intended to stay. The problem is, that this API will not compile if the list of backends is empty. This is actually very nice to have, as this is preferable over an error at runtime. However, I wanted that reviewers can compile this during review.

Reply:
That means, the `eeprom` module will not be compilable without having any real device? It is ok if there is no test case where this could happen.

Reply:
It is basically only a wrapper allowing to access multiple EEPROM devices with the same API. Without EEPROM drivers present, there is no point in having this API. So I'd say it should not compile without any backend, as this will warn users early on that a driver is missing in `USEMODULE`.

Reply:
ok